### PR TITLE
Add support for extensions containing symlinks

### PR DESCRIPTION
**Pull request by tchar**
### Link to related issue (if applicable)
#769 

### Summary of the changes in this PR
Changed [ulauncher.api.server.ExtensionDownloader.unzip](https://github.com/Ulauncher/Ulauncher/blob/90452694b6766e8bcc3235ab762d1bde1d9a770b/ulauncher/api/server/ExtensionDownloader.py#L185) to handle symlinks. As discussed in #769 ZipFile handles symlinks differently depending on the context.

This change loops over the entries in a ZipFile object and extracts them one by one. If the entry does not have symlink attributes, or it is a symlink it just proceeds, if the entry is has symlink attributes and is created as a file, it creates the symlink based on the file contents. Symlinks that point outside of the extension directory, are left as a file (original behavior).

#### Some Notes
I can create the symlink by avoiding its extraction and then reading its contents and removing it, but symlinks are rare and I wanted to keep the changes minimal. If this bugs you, I can introduce the boilerplate needed (about ~20 sloc extra) to avoid that.

### Checklist (see more [here](https://github.com/Ulauncher/Ulauncher/wiki/Code-Contribution))
- [x] Use `dev` as the base branch
- [x] Follow the [Python Code Style Guides](https://github.com/Ulauncher/Ulauncher/wiki/Python-Code-Style-Guides)
- [ ] Write unit tests for your changes when applicable
- [x] If your changes alters the behavior or introduce new functionality, please update the documentation accordingly
- [ ] All tests are passing

Just a note. There is no actual testing for `ulauncher.api.server.ExtensionDownloader.unzip`, It is mocked as far as I can tell. If you would like some tests on the unzip, I can write them, but I would need to introduce a sample zip file to be tested against (Python's ZipFile doesn't handle well symlinks so it is hard to emulate it in memory)

### Tested environment (distro, desktop environment and their versions)
- My computer
    - Linux: 5.11.0-25-generic
    - Distro: Ubuntu 20.04.2 LTS
    - DE: Gnome 3.36.8 (X11)
 - Docker image provided by ulauncher

### CODE REVIEW

**Comment by friday** [comment]
Appreciated! I hope I can get to this soon.

File: ulauncher/api/server/ExtensionDownloader.py [hunk 1]
```diff
@@ -182,7 +182,20 @@ def unzip(filename: str, ext_path: str) -> None:
     temp_ext_path = mkdtemp(prefix='ulauncher_dl_')
 
     with ZipFile(filename) as zipfile:
-        zipfile.extractall(temp_ext_path)
+        for zipinfo in zipfile.infolist():
+            filepath = zipfile.extract(zipinfo, temp_ext_path)
+            # Check if file is symlink from its attributes or zipfile did extract it as a symlink
+            # see https://learning-python.com/cgi/showcode.py?name=ziptools/ziptools/ziptools/zipsymlinks.py
+            if (zipinfo.external_attr >> 28) != 0xA or os.path.islink(filepath):
```
Comment (troycurtisjr) [inline ulauncher/api/server/ExtensionDownloader.py:189]: I see, this is due to the fact that to git, and thus inside this zip archive, the symlink is really just a text file with the path as contents. So this snippet is trying to convert that normal text file into a true symlink, as long as it is within the extension directory. Could you add a few comments calling out the fact that most/all of the times zips from Git will have the symbolic links represented as text files? That might help future us understand what is going on here and why.

File: ulauncher/api/server/ExtensionDownloader.py [hunk 2]
```diff
@@ -182,7 +182,20 @@ def unzip(filename: str, ext_path: str) -> None:
...
+                continue
+            with open(filepath) as f:
```
Comment (troycurtisjr) [inline ulauncher/api/server/ExtensionDownloader.py:191]: I might be concerned about whether or not this is a directory, but I think the symlink check above is sufficient to be sure that if we get here what we have is a small text file containing the path to point to. :+1:

**Review by troycurtisjr — CHANGES_REQUESTED** [review]
@tchar would you be up for adding a unit test for this change? I could definitely see changes being made accidentally that break the capability. Also I think it is unusual enough it might be a while before we actually figure out it got broke.

File: ulauncher/api/server/ExtensionDownloader.py (hunk 1 above)
Comment (tchar) [reply ulauncher/api/server/ExtensionDownloader.py:189]: Actually it is not only to git but to any normal  zip file. The current unzip function cannot handle any of these due to `ZipFile` not handling symlinks. You can test that by creating locally a zip with symlinks inside and pass it to Ulauncher's unzip function (that is what I actually tested in the first place). The `zipinfo.external_attr >> 28` check is generic as can be seen [here](https://learning-python.com/cgi/showcode.py?name=ziptools/ziptools/ziptools/zipsymlinks.py) at the `isSymlink` function.

I tested my function against a directory downloaded with Ulauncher's download_zip function and also a normal zip created locally by me totally unrelated to git. It works in both cases.

File: ulauncher/api/server/ExtensionDownloader.py (hunk 2 above)
Comment (tchar) [reply ulauncher/api/server/ExtensionDownloader.py:191]: Directories return False for (zipinfo.external_attr >> 28) != 0xA.

Actually my first tries had a safeguard like this
```python
# zipinfo.is_dir(), only python >= 3.6
if (zipinfo.external_attr >> 28) != 0xA or zipinfo.is_dir() or os.path.islink(filepath):  
    ...
```

**Comment by tchar** [comment]
> @tchar would you be up for adding a unit test for this change? I could definitely see changes being made accidentally that break the capability. Also I think it is unusual enough it might be a while before we actually figure out it got broke.

Yeah sure I can add tests, but I need to add a couple of zips somewhere in the `tests` directory and actually extract them (without mocker) when testing. I can do that, no problem.

I would do it in-memory but since ZipFile struggles with symlinks in the first place I can't create them in-memory using it.

**Comment by friday** [comment]
I forked your extension and reverted to the state when you had symlinks: https://github.com/friday/ulauncher-albert-calculate-anything

Then instead of downloading the zip, I downloaded the tarball: https://github.com/friday/ulauncher-albert-calculate-anything/tarball/master

I just followed the tarball documentation on how to extract:
```
import tarfile
tf = tarfile.open("friday-ulauncher-albert-calculate-anything-v0.8.0-27-g65e34de.tar.gz", mode="r")
tf.extractall()
```

This worked straight away for me, and is much cleaner imo, so I suggest that way instead. The only issue is that the root directory has the commit hash in it. `g65e34de`, so this needs to be handled instead.

**Comment by friday** [comment]
Oh actually, you can just change the file extension on the download url from zip to tar.gz and it doesn't include the commit hash. I couldn't find that documented anywhere though.

**Comment by friday** [comment]
Alternative implementation using tar #773

**Comment by tchar** [comment]
> I forked your extension and reverted to the state when you had symlinks: https://github.com/friday/ulauncher-albert-calculate-anything
> 
> Then instead of downloading the zip, I downloaded the tarball: https://github.com/friday/ulauncher-albert-calculate-anything/tarball/master
> 
> I just followed the tarball documentation on how to extract:
> 
> ```
> import tarfile
> tf = tarfile.open("friday-ulauncher-albert-calculate-anything-v0.8.0-27-g65e34de.tar.gz", mode="r")
> tf.extractall()
> ```
> 
> This worked straight away for me, and is much cleaner imo, so I suggest that way instead. The only issue is that the root directory has the commit hash in it. `g65e34de`, so this needs to be handled instead.

Pff I was hoping to get some PR to fix some of the mess in my code.

Yeah tarfile looks better, you could wait though to make the changes in this branch so we don't make multiple PRs for the same thing.

**Comment by friday** [comment]
> Yeah tarfile looks better, you could wait though to make the changes in this branch so we don't make multiple PRs for the same thing and I didnt sit and try to contribute for no reason.

Please understand that getting your contributions merged is never guaranteed. I only asked for the PR because you said you had written the code already. A PR just makes that work more formal and easy to review, but I never wanted to encourage you to put time on this. My plan was to look for an alternative to avoid introducing the extra complexity.

Most of the PRs which are on the list that haven't been merged so far will never be merged. There are many considerations involved in making these decisions that maintainers have to weight, because if it's a mess or seems risky, it'll be their mess to work with later. And if you have a long experience working with software development you do get pickier (for a good reason). It also takes a lot of time to review PRs and suggest changes, and in this case it took very little time for me to make that PR after doing the research, so it was easier for me that way.

**Comment by tchar** [comment]
> > Yeah tarfile looks better, you could wait though to make the changes in this branch so we don't make multiple PRs for the same thing and I didnt sit and try to contribute for no reason.
> 
> Please understand that getting your contributions merged is never guaranteed. I only asked for the PR because you said you had written the code already. A PR just makes that work more formal and easy to review, but I never wanted to encourage you to put time on this. My plan was to look for an alternative to avoid introducing the extra complexity.
> 
> Most of the PRs which are on the list that haven't been merged so far will never be merged. There are many considerations involved in making these decisions that maintainers have to weight, because if it's a mess or seems risky, it'll be their mess to work with later. And if you have a long experience working with software development you do get pickier (for a good reason). It also takes a lot of time to review PRs and suggest changes, and in this case it took very little time for me to make that PR after doing the research, so it was easier for me that way.

Yeah I had written the code but the PR requires from me to download the repo, branch dev, install docker and deps, format the code accordingly to the repo's specs, pass the tests etc. The prototype I had did not need any of that work. And most of the work was basically configuring my environment with linting deps etc for the PR, not writing the code.

I am just pointing out that not making comments on the existing PR, fixing it in the maintainer's way, discourages people from spending any time contributing to any project. The work is not the code, that's easy, the work is making the code ready for review etc.

I am not arguing on the pickyness actually this is correct but it can be done through discussion and code review.

Anyways. You can close this PR, since you already fixed it.

**Comment by friday** [comment]
This happened to me many times @tchar. For example in https://github.com/tkashkin/GameHub/pull/171. I had to make the PR to prove it could be done, and hopefully it helped the maintainer make his counter-suggestion. The solution he went with was better, so I was happy that my use case was covered, and I appreciate that he did that instead of telling me how to change my code to become his code. That back and forth would have been a waste of his time and mine.

The experience definitely didn't encourage me to continue contributing to that project, but I blame my PR and limited knowledge of Vala async event handling, and not the project maintainer for doing the right thing.

**Comment by tchar** [comment]
@friday 

I partly agree with you.

Tarfile is indeed a better solution if you don't care changing dependencies.

However given that you are a project maintainer and you have experience with softeng I thought you had a reason to keep `ZipFile` and already knew that this problem exists. I simply went with ZipFile because you said you don't want to alter dependencies. And this is just experience with talking with multiple project maintainers because one doesn't like x and the other doesn't like y.

Anyways my main problem with these kind of projects is that I have to maintain a separate branch for a ton of projects because [either a feature is not implemented](https://github.com/Ulauncher/Ulauncher/issues/273) and I have to [implement](https://github.com/Ulauncher/Ulauncher/compare/dev...tchar:requirements?expand=1) it myself or there are [memory issues causing  a lightweight program to be as heavy as gnome](https://github.com/tchar/qtile/commit/4127cf35bcce2a989647cfade159c1571edf6c51) or just simply devs don't want to support a feature.

**Comment by friday** [comment]
ZipFile and TarFile are both standard Python modules written in Python with no external dependencies. Switching from one to the other doesn't change dependencies. Switching to use git or to use the `unzip` or `tar` commands via subprocess changes the dependencies, because then everyone using Ulauncher has to install those. Although most users probably have them, it's better to avoid adding dependencies.

The rest is good criticism imo. I also want to improve things like this. Extension dependencies is one of the more important issues for Ulauncher to handle imo, along with HiDPI support and fixing crashes. We are prioritizing this. This is a big and important internal change and something I want to fully understand before I integrate it.

**Pull request closed without merging**